### PR TITLE
stdlib-shims: fails to build on 4.12.0+trunk

### PR DESCRIPTION
**Pull request by avsm**
upstream fix proposed at https://github.com/ocaml/stdlib-shims/pull/17

### CODE REVIEW

**Comment by dra27** [comment]
This was fixed in Dune in https://github.com/ocaml/dune/pull/3576

**Comment by avsm** [comment]
Sounds like we need to conflict this with dune < 2.7 and ocaml <4.12 then

**Comment by avsm** [comment]
does https://github.com/ocaml/opam-repository/pull/16983/commits/241ff77171fde471ef7951c27bd6ec27199a5a3a look right @dra27?

File: packages/stdlib-shims/stdlib-shims.0.1.0/opam [hunk 1]
```diff
@@ -8,8 +8,9 @@ bug-reports: "https://github.com/ocaml/stdlib-shims/issues"
 tags: ["stdlib" "compatibility" "org:ocaml"]
 license: ["typeof OCaml system"]
 depends: [
+  "ocaml" {>="4.02.3"}
   "dune"
-  "ocaml" {>= "4.02.3"}
+ ("dune" {> "2.7.0"} | "ocaml" {<"4.12.0~~"})
```
Comment (kit-ty-kate) [inline packages/stdlib-shims/stdlib-shims.0.1.0/opam:13]: why `>` instead of `>=`? I don't think dune 2.7.0 has been released yet.

Comment (avsm) [reply packages/stdlib-shims/stdlib-shims.0.1.0/opam:13]: it should be `>=` indeed! good catch

Comment (avsm) [inline packages/stdlib-shims/stdlib-shims.0.1.0/opam:13]:
```suggestion
 ("dune" {>= "2.7.0"} | "ocaml" {<"4.12.0~~"})
```

**Comment by hannesm** [comment]
I'm curious whether dune < 2.7.0 should require ocaml < 4.12.0, and not bother with constraining stdlib-shims at all... since it looks to me that more packages (all with empty modules) will encounter this issue...

File: packages/stdlib-shims/stdlib-shims.0.1.0/opam [hunk 2]
```diff
@@ -8,8 +8,9 @@ bug-reports: "https://github.com/ocaml/stdlib-shims/issues"
 tags: ["stdlib" "compatibility" "org:ocaml"]
 license: ["typeof OCaml system"]
 depends: [
+  "ocaml" {>="4.02.3"}
   "dune"
-  "ocaml" {>= "4.02.3"}
+ ("dune" {>= "2.7.0"} | "ocaml" {<"4.12.0~~"})
```
Comment (dra27) [inline packages/stdlib-shims/stdlib-shims.0.1.0/opam:13]: Sorry, I should not answer on my phone at the weekend. This isn't right!
```suggestion
 ("dune" {>= "2.7.0"} | "dune" & "ocaml" {<"4.12.0~~"})
```

**Comment by dra27** [comment]
@hannesm is correct! Are you aware of any packages which would be affected by this change, though - it's not empty modules, but empty _libraries_ (i.e. a library with no modules)

**Comment by avsm** [comment]
We'll find this out soon enough when we start doing bulk tests on 4.12. I suspect the current restriction is fine, and we can apply the big dune restriction if there turn out to be a non-trivial number of affected packages.

**Comment by mseri** [comment]
Should we merge this or you prefer to wait? In the end it is a small change to keep the issue under control, it seems a reasonable start

**Comment by avsm** [comment]
Agreed. I'll merge go ahead and merge this, and it can be tested with #17005

**Pull request merged**